### DISCUSSION TITLE
chapter_dynamic_programming/intro_to_dynamic_programming/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#573] [Announcements]
# chapter_dynamic_programming/intro_to_dynamic_programming/

动画图解、一键运行的数据结构与算法教程

https://www.hello-algo.com/chapter_dynamic_programming/intro_to_dynamic_programming/

**Comment by zhuasdfg** [comment]
牛逼，向上递归变向下递归，再优化递归的剪枝，指数变线性，递归再优化成向上，最后一个循环两个指针就解决问题了，太强大了。

**Reply by 1669119595** [reply]
这个递进过程看着是真的舒服

**Reply by RockyZhang4031** [reply]
看起来非常可口

**Comment by i-Remus** [comment]
这一章是真的看懂了，真tm好！！！感谢 K神让我学会了一种算法

**Comment by Donglin-SHI** [comment]
```C
int climbingStairsDPComp(int n)
{
    int a = 1, b = 2;

    while (--n)
        b += a, a = b - a;

    return a;
}
```
这个不知能否作为14.1.4的C语言实现。

**Comment by hpp0716** [comment]
而你，我的朋友，你才是真正的英雄

**Comment by memorycancel** [comment]
膜拜！

**Comment by PhantomMaa** [comment]
使用 res[0] 记录方案数量。
这里感觉可以替换为AtomicInteger，更好理解

**Reply by krahets** [reply]
可以哇， `res[0]` 适用于所有语言，所以用了这个

**Reply by Meta-YZ** [reply]
请问res为什么是个数组才能运行正确结果，直接res = 0 然后res += 1为啥不行呢？

**Reply by krahets** [reply]
@Meta-YZ 使用 `res = [0]` ，由于列表是可变的，这样可以确保在递归过程中对方案数量的累加是对同一个对象的修改，从而正确地计算出总的方案数量。如果只用一个整数 `res = 0` ，则每个递归函数中的 `res` 都是独立的，无法在递归调用中累加方案数量。

**Reply by Asce-ops** [reply]
这里我也想了很久，我理解应该是这样的：如果用`res = 0`，递归过程中后入栈的函数空间确实可以使用先入栈的函数空间中的 res 的结果，但是因为 int 是不可变的类型，这样后入栈的函数空间中 res 被更新了但这个更新没办法同步到先入栈的函数空间，结果就是一旦入栈对应函数空间中的 res 就不会再变化了

**Comment by ICUSBZ** [comment]
我用排列组合做的还有救吗，程序里面没有一丁点算法思维，纯纯计算阶乘，难顶

**Reply by krahets** [reply]
有救啊，慢慢培养算法思维，加油！

**Comment by newcakes** [comment]
太细了，庖丁解牛，看的太爽了。

**Comment by QK51643729** [comment]
这里第一个回溯算法存在问题，需要保证choices中递增排序才能使用break剪枝

**Reply by krahets** [reply]
很细心奥！用 continue 剪枝更严谨些

**Reply by diplomatist** [reply]
perfect!

**Reply by sunshinesDL** [reply]
借下楼😁，第一个回溯程序的 backtrack() 函数，源代码的返回值写的是 int型，但实际是没有返回值的，
不过感谢 K大，这启发了我尝试不把 res 定义为一个列表，而是将其定义为整数，并且在 backtrack() 中显示传递 res 的值，以在递归函数结束时成功把对 res 的修改传递回上层函数。代码如下：
```python
def backtrack(choices: list[int], state: int, n: int, res: int) -> int:
    """回溯"""
    # 当爬到第 n 阶时，方案数量加 1
    if state == n:
        res += 1
        return res
    # 遍历所有选择
    for choice in choices:
        # 剪枝：不允许越过第 n 阶
        if state + choice > n:
            continue
        # 尝试：做出选择，更新状态
        res = backtrack(choices, state + choice, n, res)
        # 回退
    return res


def climbing_stairs_backtrack(n: int) -> int:
    """爬楼梯：回溯"""
    choices = [1, 2]  # 可选择向上爬 1 阶或 2 阶
    state = 0  # 从第 0 阶开始爬
    res = 0  
    res = backtrack(choices, state, n, res)
    return res


"""Driver Code"""
if __name__ == "__main__":
    n = 9

    res = climbing_stairs_backtrack(n)
    print(f"爬 {n} 阶楼梯共有 {res} 种方案")
```

**Comment by peng-yq** [comment]
动态规划的解法在没有空间优化前从时间复杂度和空间复杂度的角度和记忆化搜索是性能相当的，但记忆化搜索涉及函数栈的增长，极端情况下可能导致栈溢出。（不知道我说的对不，感觉可以增加这个？）

**Reply by krahets** [reply]
对的，使用迭代和递归实现相同的任务时，迭代一般比递归更加高效，因为递归涉及到调用函数产生的额外开销。

**Reply by ShiinaMahiru-repo** [reply]
感觉不会吧， 不会增加多少， 那个mem数组相对于dfs是全局的 所以对栈也不会 有太大影响吧 感觉 只是回溯时 多了个赋值操作 我是这么理解的 不知道对不对，回一下我

**Reply by krahets** [reply]
@ShiinaMahiru-repo Hi，这里指的不是 mem 数组带来的内存占用，而是记忆化搜索在递归中产生的栈帧空间占用。可以看下“图 14-4 记忆化搜索对应递归树”

**Comment by simonwater** [comment]
斐波那契数的求解。。

**Comment by 08822407d** [comment]
得，研究了半天发现是斐波那契数列的递归和循环两种解法

**Reply by UniKylin** [reply]
有木有可能是一个解法，两道题！

**Comment by gongluck** [comment]
循序渐进👍

**Comment by wwangxinhao** [comment]
简洁明了，写的非常好！

**Comment by DoYangTan** [comment]
写得太好了！

**Comment by xuejian-git** [comment]
质量太高了吧

**Comment by ShiinaMahiru-repo** [comment]
说起来这个滚动是不是可以这样  
1. b = a + b;
```cpp
    a = b - a;
```
2. tmp = a;
```cpp
    a = b;
    b = a + tmp;
```
3. tmp = b;
```cpp
    b = a + b;
    a = tmp;
```

**Reply by krahets** [reply]
2., 3. 都是正确写法。不推荐 1.，可读性不好，且考虑通常比 2.,3. 更低。1. 适合做脑筋急转弯题

**Reply by ShiinaMahiru-repo** [reply]
哦， 明白了， 谢谢回复

**Reply by ShiinaMahiru-repo** [reply]
让我想起了滚动数组

**Reply by syutengu** [reply]
手欠忍不住这样写：
```js
    let a,b
    for (let i = 3; i <= n; i++) [b=2, a=1] = [a + b, b]
```

**Reply by syutengu** [reply]
然后发觉是错的

**Reply by syutengu** [reply]
最多只能耍到这里了，不知道还有更好玩的嘛。
```js
    let [a, b] = [1, 2]
    for (let i = 3; i <= n; i++) [b, a] = [a + b, b] //or b=[a+b,a=b][0]
```

**Comment by ahui-5144** [comment]
```java
climbing_stairs_backtrack.java
        if (state == n) {
            res.set(0, res.get(0) + 1);
            return;
        }
```
我试了几个数据,这里加上return后结果没有改变,而且运行速度更快,请问这样子加上return会有什么隐患嘛(作者这里没有加return是有什么考虑嘛)

**Reply by krahets** [reply]
Hi，这样做主要是为了符合文中的描述“每当到达楼梯顶部时就将方案数量加 1 ，当越过楼梯顶部时就将其剪枝”。
针对这个 case ，加 `return` 会变的更快（提前返回），没有什么隐患

**Comment by coder-yangxin** [comment]
## 举个手🤔

> _爬楼梯_ 回溯算法实现中为什么不需要回退操作，这个案例和13章中需要回退的案例有什么区别呢

**Reply by hpstory** [reply]
Hi @coder-yangxin , 其中一个区别就是本章的代码求的是方案数，而不是所有方案。调用`Backtrack`之后不需要关心后面的情况，最终当爬到第 n 阶时，方案数量就会变化。而如果需要求具体的方案，比如将`choice`放入到方案中，`Backtrack`递归完成后，当前的`choice`已经在方案的列表中，这个时候需要把他拿出来，下轮循环再放入其他`choice`。

**Reply by krahets** [reply]
Hi，因为在爬楼梯问题中，状态 `state` 的类型是 int ，我们在调用函数尝试（做出决策、更新状态）时，是将它值传递进去的。每个开启的函数中的 `state` 都是“独立”的，因此无须回退操作（恢复状态）。

**Reply by coder-yangxin** [reply]
明白了！

>回溯中利用 **递归** 完成 **前进（递）** 和 **后退（归）** 操作，以此往方案中 **添加** 或 **移除** 选择项（对应框架代码中的 `尝试` 和 `回退` 操作）。

本案例中不记录选择项 并且 上一轮迭代中有自己的的state，所以 **后退（归）** 时 移除选择项（恢复状态）操作是不必要的

**Reply by likz1994** [reply]
如果，你的写法是：
```cpp
	state = state + choices；
	// 尝试：做出选择，更新状态
    backtrack(choices, state, n, res);
    // 回退（就需要进行 state的 回退操作）
	state = state - choices；
```

**Comment by swaggywilliam** [comment]
断断续续学了动态规划一个月了，反复看了三四遍，今天终于感觉通了一些

**Comment by CHLK** [comment]
大佬牛逼，层层递进，写的很明白，看了这么多解析动态规划的，还是本文最易懂

**Comment by macj2010** [comment]
这个问题莫名有点像上一章《回溯》里面的《子集》问题，给定一个集合[3,4,5]，求他们和是9的子集有多少种。这个问题换成集合[1,2]，和是n的子集。不同于前面要考虑去重，这里不用考虑。{1，2}、{2，1}被认为是不同的结果。那我很好奇，当然dfs也很不错，但是如果上楼梯的步数变成[1,2,3]，是不是就变成dfs(i)=dfs(i-1)+dfs(i-2)+dfs(i-3)。

**Reply by hpstory** [reply]
Hi @macj2010 , 是的会变成`dfs(i)=dfs(i-1)+dfs(i-2)+dfs(i-3)`，动态规划和记忆化搜索是可以相互转换的，回溯的话包含很多重复计算。

**Comment by MuZiYuHui** [comment]
牛啊，循循渐进，先从回溯法切入问题，再一步一步优化，对我这个初学者来说很好理解

**Comment by Vstornzw** [comment]
C++
是否可以增加isDirty,写成：
```cpp
/* 回溯 */
void backtrack(vector<int> &choices, int state, int n, vector<int> &res) {
    // 当爬到第 n 阶时，方案数量加 1
    bool isDirty = false;
    if (state == n) {
        res[0]++;
        isDirty = true;
    }
    // 遍历所有选择
    if (!isDirty) {
        for (auto &choice : choices) {
            // 剪枝：不允许越过第 n 阶
            if (state + choice > n)
                continue;
            // 尝试：做出选择，更新状态
            backtrack(choices, state + choice, n, res);
            // 回退
        }
    }
}
```

**Reply by krahets** [reply]
这样做不太有必要，直接在记录解后 return 就可以了

**Reply by sishugang** [reply]
我看第一段代码没有return，是没有return这个函数也可以结束吗？脑子有点糊了

**Comment by ghost** [comment]
妙啊！！，真的是太妙了，大哥哥，我是小学生，tql

**Reply by krahets** [reply]
哈哈，你真的是小学生吗？

**Comment by haerker** [comment]
k神，看到这里突然对动态规划和分治有些模糊，都是将大问题拆分为小问题。我的理解：分治拆分的小问题是独立的，不重复的，未知的，所以需要有“合并”的操作，所以需要递归。而动态规划的小问题具有重复性，已知的，所以可以采用迭代。这样理解正确吗？

**Reply by shirogome** [reply]
正确的。动态规划的特征是Overlapping Subproblems和Optimal Substructure。

**Comment by ovo-Tim** [comment]
好像还没人发现，14.1.4 Rust 代码可以用解构省去 tmp 变量: `(a, b) = (b, a+b);`
``` rust
/* 爬楼梯：空间优化后的动态规划 */
fn climbing_stairs_dp_comp(n: usize) -> i32 {
    if n == 1 || n == 2 {
        return n as i32;
    }
    let (mut a, mut b) = (1, 2);
    for _ in 3..=n {
        (a, b) = (b, a+b);
    }
    b
}
```

**Reply by RocGit** [reply]
C#9.0以后 也可以写成 (a, b) = (b, a+b); 的形式

**Reply by xbsheng** [reply]
js/ts版本也可以这样写
```ts
/* 爬楼梯：空间优化后的动态规划 */
function climbingStairsDPComp(n: number): number {
  if (n === 1 || n === 2) return n
  let a = 1,
    b = 2
  for (let i = 3; i <= n; i++) {
    ;[a, b] = [b, a + b]
  }
  return b
}
```

**Comment by moonpoet** [comment]
写的真的很棒，层层递进，逐层深入，看下来没有压力，点赞

**Comment by jieqiyue** [comment]
打个卡666

**Comment by selfboot** [comment]
做了个 [动态规划硬币找零的可视化](https://gallery.selfboot.cn/en/algorithms/dpcoin)
欢迎来可视化页面直观感受

**Reply by selfboot** [reply]
https://github.com/user-attachments/assets/534dee4d-fce2-4742-afe9-03d8794058cc

**Reply by akfc58** [reply]
提一个建议：例如1，3，4，5组成7时，显示的绿框（采取的最优方案）在有几个局部最优方案时，框在最后一个上面更好理解（保留最后一个局部最优解），目前是框在第一个。

**Reply by selfboot** [reply]
> 提一个建议：例如1，3，4，5组成7时，显示的绿框（采取的最优方案）在有几个局部最优方案时，框在最后一个上面更好理解（保留最后一个局部最优解），目前是框在第一个。

这里为啥在最后一个会更好理解些呀？

**Reply by apricity-6** [reply]
我个人觉得硬币找零问题用动态规划的话应该解决有多少种方法能够找零的问题，如果只是最少硬币数，从直觉上来说就是金额越大越好，直接拿要找的金额 除以 当前可找零的最大面值硬币，商是硬币数，余数接着找可找零的最大面值硬币。递归或者循环应该更好理解一点。每次最大可找零面值还可以通过最大堆来优化处理。只是找最少硬币数很难理解使用动态规划的好处和必要性。

**Reply by apricity-6** [reply]
我尝试使用最大堆和贪心算法去解决硬币问题， 常见硬币是可以完成的，但不支持自定义硬币面额，如果面额是[1,3,4]，金额是6，得出的结论是 4+1+1 3个硬币，实际可以3+3 两个硬币完成。如果自定义硬币面额，动态规划还是有必要的。

**Comment by syutengu** [comment]
记忆化搜索，js版以下写法会不会更好。浏览器环境下，fn(31415)还能出结果，但原文写法的话已经too much recursion了。
```js
  /* 记忆化搜索 */
  function climb_stairs_dfs_mem(n) {
    let counter = 0
    const mem = [0, 1, 2]
    const res = dfs(n)
    console.log(counter)
    return res

    function dfs(n) {
      counter++
      if (n === 1 || n === 2) return n
      mem[n - 2] ??= dfs(n - 2)
      mem[n - 1] ??= dfs(n - 1)
      return mem[n - 1] + mem[n - 2]
    }
  }
```

**Reply by syutengu** [reply]
读到第三节，对比一下fn(314159)，马上就理解动态规划反直觉逆解问题以消除递归的优势所在了！

**Reply by syutengu** [reply]
或者说递归才是反直觉的，动态规划把它拨乱反正，负负得正了。

**Reply by aleyx1979** [reply]
我倒是觉得递归才是直觉的——从把“数学定义”转换到“代码”的角度看。
我观察到一些搞科研的朋友能够很快的接受“递归”，反而在 “for index in range(len(xxx))” 这样的循环、多重循环等“简单问题”上纠结很久 [:dodge]

**Reply by syutengu** [reply]
是的，写代码的时候也是这个感觉。
递归是“由条件指向答案”（from - to），
而DP是“答案来自前答案（最终来自条件）”(to - from)。

**Comment by TanZiDong** [comment]
请问能给函数的参数加注释吗？

**Comment by YiChau** [comment]
LeetCode 爬楼梯 传送门：https://leetcode.cn/problems/climbing-stairs/

**Reply by jiuerya7** [reply]
看我动态规划苗姐🤓

**Comment by hju-wang** [comment]
我觉的在  `if state == n {}`加上return 也是普遍成立的，因为我已经到达了目标，再去尝试任何解都会触发剪枝头，因为我的choice 是`{1,2}`

**Comment by jiuerya7** [comment]
发明动态规划的人真是个天才！

**Reply by yyym2017** [reply]
不是所谓的发明，只能叫发现，规律本身存在于大自然

**Comment by rainnnnnnnnnnnnnnnnnnnnnn** [comment]
分治的本质是自顶而下，将问题拆解成更小的子问题，直到子问题可以很容易地解决。
回溯则像平行地探索所有种类的可能性，某种程度上等同于DFS（深度优先搜索），尽管回溯也有用到递归，但是它的中心思想却不是分治。
而动态规划的本质是自下而上，由最小子问题的解（初试状态）+ 父问题与子问题之间的关系 （状态转移方程）不断向上循环，直到得到所求 问题的解。

**Reply by zhuyihuan68** [reply]
分治是子问题之间互不影响 拆完合并起来 就是原问题的解
回溯就是通过 循环迭代 或者 递归 穷举所有可能性 在剪枝
动态规划自下而上 将原来可能用递归的问题 现在可以通过循环解决

**Comment by superfeimao** [comment]
《细心的读者可能发现了》

**Comment by fanyu0122** [comment]
爬到第 i-1 阶的方案数加上爬到第 i-2 阶的方案数就等于爬到第 i 阶的方案数。

**Comment by fanyu0122** [comment]
为什么：
由于每轮只能上 1 阶或 2 阶，因此当我们站在第 i 阶楼梯上时，上一轮只可能站在第 i-1 阶或第 i-2 阶上。换句话说，我们只能从第 i-1 阶或第 i-2 阶迈向第 i 阶。

所以：
爬到第 i-1 阶的方案数加上爬到第 i-2 阶的方案数就等于爬到第 i 阶的方案数。

**Comment by alahenell** [comment]
你这个感受 **非常有道理，而且非常“到位”**。
我可以直接说一句判断：

> **从回溯到 DP，真正难的不是公式，而是“状态语义的转变”**
> 你说的这个点，正是 80% 人卡住的地方。

下面我分层帮你把这个“跨越点”讲透。

---

## 一、你现在的困惑本质是什么？

你说的是爬楼梯这个例子，对吧。

### 回溯时你脑子里是这样的：

```text
state = 现在站在第几阶
choice = {+1, +2}
res = 所有路径的数量
```

也就是：

> **状态 = 过程中的位置**
>
> **答案 = 叶子节点的计数**

这是**搜索视角**。

---

## 二、DP 的“思维断层”到底断在哪里？

DP 要你突然改成：

```text
dp[i] = 到第 i 阶的方案数
```

这一步让人非常不适，因为：

* 在回溯里
  👉「方案数」是**全局统计结果**
* 在 DP 里
  👉「方案数」成了**状态本身**

你感觉像是：

> “我还没走完，怎么就已经知道方案数了？”

这是非常自然的心理冲突。

---

## 三、关键认知转折（这是核心）

### 🔑 真正的转折不是「递归 → 循环」

而是：

> **从「枚举路径」转向「合并同类子问题」**

我们用一句话对比：

| 思维方式 | 在干什么                 |
| ---- | -------------------- |
| 回溯   | 我一条一条路径走完            |
| DP   | 我不关心路径长什么样，只关心“有多少条” |

---

## 四、为什么 DP 的状态必须是“方案数”？

我们把你的回溯代码抽象一下：

```python
def f(i):
    if i == n:
        return 1
    return f(i+1) + f(i+2)
```

你注意到没有：

👉 **这个函数返回的已经不是“位置”，而是“数量”**

也就是说：

> 回溯的“过程状态”
> ⬇
> 递归函数的“返回值”
> ⬇
> DP 的状态含义

---

## 五、用一条“认知桥梁”帮你跨过去

你可以这样在脑子里完成转化：

### Step 1：回溯视角（你熟的）

```text
“我站在第 i 阶，往上走，能走出多少条路？”
```

### Step 2：递归函数含义

```text
f(i) = 从第 i 阶走到 n 阶的方案数
```

⚠️ 注意：
这里 **f(i) 已经不是“位置”了，而是“数量”**

---

### Step 3：DP 状态定义

```text
dp[i] = 从第 i 阶走到 n 阶的方案数
```

或者反过来：

```text
dp[i] = 到第 i 阶的方案数
```

👉 **DP 只是把“递归返回值”存了下来**

---

## 六、为什么很多人卡在这一步？

因为他们：

* 在回溯阶段
  把「状态」理解成**世界的客观状态**
* 在 DP 阶段
  却要求把「状态」理解成**问题的答案碎片**

这是一次**语义升级**，不是技巧升级。

---

## 七、你这个总结本身已经是“DP 思维雏形”

你说的这句话：

> *“思维不能停留在回溯时状态表示台阶数，而是转到 dp 这里状态表示方案数”*

我帮你精炼成一句**教科书级总结**：

> **回溯中状态描述“我在哪”；
> DP 中状态描述“答案的一部分”。**

这句话你可以反复用在：

* 爬楼梯
* 不同路径
* 子集问题
* 背包
* 编辑距离

---

## 八、一个判断标准（给你自检用）

你可以用这句话检查自己是否真正“进入 DP 思维”：

> **如果我不关心每条路径长什么样，只关心数量/最优值，那就该用 DP**

一旦你自然地问出：

> “从这个状态出发，**结果是多少**？”

你已经在用 DP 了。

---

## 九、最后一句实话

你已经不是“不会 DP”，
你只是正在经历**从「过程思维」到「结果思维」的转型期**。

这个阶段，**感到别扭是正常现象，而且是好现象**。

如果你愿意，下一步我可以帮你做一件事：

> **把任意一个你熟的回溯问题，手把手改写成 DP**
> 不换题，只换思维。

你现在这个问题，问得非常专业。

**Comment by XiHoZi** [comment]
```python
def climbing_stairs_dp_comp(n: int) -> int:
    """爬楼梯：空间优化后的动态规划"""
    if n == 1 or n == 2:
        return n
    a, b = 1, 2
    for _ in range(3, n + 1):
        a, b = b, a + b
    return b

```
如此优雅的代码

**Comment by hgc-qwe** [comment]
如果这个在线观看有笔记功能就好了

**Comment by zhangchengxin0927-lang** [comment]
好像斐波那契

**Reply by SSZero-670** [reply]
就是斐波那契，第i项等于前两项的和
